File: membership_card_verifier/packages/verifier/test/fixtures.ts
/**
 * Shared crypto fixtures for stage tests.
 * All keys are generated once and reused across tests using the same seed patterns.
 */
import { ml_dsa44 } from "@noble/post-quantum/ml-dsa.js";
import { createCipheriv, randomBytes } from "node:crypto";
import { canonicalize } from "../src/canonicalize.js";
import { keccak256, hkdfSha3256 } from "../src/crypto.js";
import type { CardDocument, SubCardDocument } from "../src/types.js";

export function generateKeypair() {
  const { secretKey, publicKey } = ml_dsa44.keygen();
  const address = keccak256(publicKey);
  return { secretKey, publicKey, address };
}

export function sign(secretKey: Uint8Array, data: unknown): string {
  const bytes = canonicalize(data);
  const sig = ml_dsa44.sign(bytes, secretKey);
  return Buffer.from(sig).toString("base64url");
}

export function encryptForCard(pubkey: Uint8Array, plaintext: Uint8Array): Uint8Array {
  const contentKey = hkdfSha3256(pubkey, "card-content-v1");
  const nonce = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", contentKey, nonce);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return new Uint8Array(Buffer.concat([nonce, ciphertext, tag]));
}

export function makeCardDoc(
  recipientPubkey: Uint8Array,
  issuerSk: Uint8Array,
  holderSk: Uint8Array,
  pressSk: Uint8Array,
  ancestryPubkeys: string[] = [],
  extra: Record<string, unknown> = {}
): CardDocument {
  const recipientPubkeyB64 = Buffer.from(recipientPubkey).toString("base64url");
  const { publicKey: issuerPub } = ml_dsa44.keygen(); // fake issuer for CID references
  const { publicKey: pressPub } = ml_dsa44.keygen();

  const offer: Partial<CardDocument> = {
    policy_id: "QmFakePolicyCID",
    issuer_card: keccak256(new Uint8Array(issuerPub)),
    press_card: keccak256(new Uint8Array(pressPub)),
    protocol_version: "0.1",
    issued_at: "2026-06-20T00:00:00Z",
    ancestry_pubkeys: ancestryPubkeys,
    ...extra,
  };

  const issuerSigInput = { ...offer };
  const issuerSig = sign(issuerSk, issuerSigInput);

  const holderSigInput = { ...offer, issuer_signature: issuerSig, recipient_pubkey: recipientPubkeyB64 };
  const holderSig = sign(holderSk, holderSigInput);

  const pressSigInput = { ...holderSigInput, holder_signature: holderSig };
  const pressSig = sign(pressSk, pressSigInput);

  return {
    ...pressSigInput,
    press_signature: pressSig,
    issuer_signature: issuerSig,
    holder_signature: holderSig,
  } as CardDocument;
}

export function makeSubCardDoc(
  holderPubkey: Uint8Array,
  holderSk: Uint8Array,
  appPubkey: Uint8Array,
  appSk: Uint8Array,
  recipientPubkey: Uint8Array
): SubCardDocument {
  const holderAddress = keccak256(holderPubkey);
  const appAddress = keccak256(appPubkey);

  const base = {
    holder_primary_card: holderAddress,
    holder_primary_card_pubkey: Buffer.from(holderPubkey).toString("base64url"),
    app_card: appAddress,
    app_card_pubkey: Buffer.from(appPubkey).toString("base64url"),
    capabilities: ["note"],
    recipient_pubkey: Buffer.from(recipientPubkey).toString("base64url"),
    issued_at: "2026-06-20T00:00:00Z",
    attestation_level: "T2" as const,
  };

  const appSig = sign(appSk, base);
  const holderSigInput = { ...base, app_signature: appSig };
  const holderSig = sign(holderSk, holderSigInput);

  return { ...holderSigInput, holder_signature: holderSig };
}

File: membership_card_verifier/packages/verifier-py/tests/fixtures.py
"""Shared crypto fixtures for stage tests."""

import base64
import os
from dataclasses import dataclass
from typing import Any

from cryptography.hazmat.primitives.asymmetric import mldsa
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from membership_card_verifier.canonicalize import canonicalize
from membership_card_verifier.crypto import hkdf_sha3_256, keccak256


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


@dataclass
class Keypair:
    secret_key: mldsa.MLDSA44PrivateKey
    public_key: bytes
    address: str


def generate_keypair() -> Keypair:
    priv = mldsa.MLDSA44PrivateKey.generate()
    pub = priv.public_key()
    public_key = pub.public_bytes_raw()
    address = keccak256(public_key)
    return Keypair(secret_key=priv, public_key=public_key, address=address)


def sign(secret_key: mldsa.MLDSA44PrivateKey, data: Any) -> str:
    data_bytes = canonicalize(data)
    signature = secret_key.sign(data_bytes)
    return _b64url_encode(signature)


def encrypt_for_card(pubkey: bytes, plaintext: bytes) -> bytes:
    content_key = hkdf_sha3_256(pubkey, "card-content-v1")
    nonce = os.urandom(12)
    ciphertext_and_tag = AESGCM(content_key).encrypt(nonce, plaintext, None)
    return nonce + ciphertext_and_tag


def make_card_doc(
    recipient_pubkey: bytes,
    issuer_sk: mldsa.MLDSA44PrivateKey,
    holder_sk: mldsa.MLDSA44PrivateKey,
    press_sk: mldsa.MLDSA44PrivateKey,
    ancestry_pubkeys: list[str] | None = None,
    extra: dict[str, Any] | None = None,
) -> dict[str, Any]:
    if ancestry_pubkeys is None:
        ancestry_pubkeys = []
    if extra is None:
        extra = {}

    recipient_pubkey_b64 = _b64url_encode(recipient_pubkey)
    issuer_pub = mldsa.MLDSA44PrivateKey.generate().public_key().public_bytes_raw()
    press_pub = mldsa.MLDSA44PrivateKey.generate().public_key().public_bytes_raw()

    offer: dict[str, Any] = {
        "policy_id": "QmFakePolicyCID",
        "issuer_card": keccak256(issuer_pub),
        "press_card": keccak256(press_pub),
        "protocol_version": "0.1",
        "issued_at": "2026-06-20T00:00:00Z",
        "ancestry_pubkeys": ancestry_pubkeys,
        **extra,
    }

    issuer_sig_input = dict(offer)
    issuer_sig = sign(issuer_sk, issuer_sig_input)

    holder_sig_input = {
        **offer,
        "issuer_signature": issuer_sig,
        "recipient_pubkey": recipient_pubkey_b64,
    }
    holder_sig = sign(holder_sk, holder_sig_input)

    press_sig_input = {**holder_sig_input, "holder_signature": holder_sig}
    press_sig = sign(press_sk, press_sig_input)

    return {
        **press_sig_input,
        "press_signature": press_sig,
        "issuer_signature": issuer_sig,
        "holder_signature": holder_sig,
    }


def make_sub_card_doc(
    holder_pubkey: bytes,
    holder_sk: mldsa.MLDSA44PrivateKey,
    app_pubkey: bytes,
    app_sk: mldsa.MLDSA44PrivateKey,
    recipient_pubkey: bytes,
) -> dict[str, Any]:
    holder_address = keccak256(holder_pubkey)
    app_address = keccak256(app_pubkey)

    base: dict[str, Any] = {
        "holder_primary_card": holder_address,
        "holder_primary_card_pubkey": _b64url_encode(holder_pubkey),
        "app_card": app_address,
        "app_card_pubkey": _b64url_encode(app_pubkey),
        "capabilities": ["note"],
        "recipient_pubkey": _b64url_encode(recipient_pubkey),
        "issued_at": "2026-06-20T00:00:00Z",
        "attestation_level": "T2",
    }

    app_sig = sign(app_sk, base)
    holder_sig_input = {**base, "app_signature": app_sig}
    holder_sig = sign(holder_sk, holder_sig_input)

    return {**holder_sig_input, "holder_signature": holder_sig}
